amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShutdownTask.java

```diff
@@ -116,25 +117,37 @@ public TaskResult call() {
             try {
                 log.debug("Invoking shutdown() for shard {} with childShards {}, concurrencyToken {}. Shutdown reason: {}",
                         leaseKeyProvider.apply(shardInfo), childShards, shardInfo.concurrencyToken(), reason);
+                ShutdownReason localReason = reason;
 
                 final long startTime = System.currentTimeMillis();
-                if (reason == ShutdownReason.SHARD_END) {
+                final Lease currentShardLease = leaseCoordinator.getCurrentlyHeldLease(leaseKeyProvider.apply(shardInfo));
+                if (localReason == ShutdownReason.SHARD_END) {
```

Comment:
Can we wrap 124 to 147 in a function `localReason = attemptPersistingChildShardInfoAndOverrideShutdownReasonOnFailure()`

Reply:
Addressed